### ISSUE TITLE
Instant Run makes Moshi 10X slower

### ISSUE THREAD

**Issue by thierryd** [square/moshi#208]
This issue started from a discussion with @swankjesse located here: https://twitter.com/eldimo/status/787992388308463617

The gist is at https://gist.github.com/thierryd/6c9ddf174324e24c5c5fd2aaa1e2aa40
---
I have benchmarked Moshi against GSON (https://github.com/google/gson) and found Moshi 10X slower than GSON. @swankjesse did some test and found out that whitespace could be the culprit (see https://github.com/square/moshi/issues/207). While I was validating his finding, I ran my benchmark on another machine and found out similar speed between Moshi and GSON. I then realized that Instant Run was turned off on that machine. 

This also explain why I was seeing worse performance on my Nexus 9 with 7.0 than my Nexus 10 with 5.0. Instant Run is not enabled when you run on a device with 5.0.
---
**tl;dr** Enabling Instant Run makes Moshi 10x slower than GSON

**Comment by swankjesse** [comment]
Strange. I can’t think of anything that’d make Gson and Moshi exercise different paths in the compiler.

**Comment by thierryd** [comment]
@swankjesse Do you think this should be filled on b.android.com instead?

**Comment by swankjesse** [comment]
Yeah, good idea.

**Comment by thierryd** [comment]
@swankjesse Done. See https://code.google.com/p/android/issues/detail?id=225743

**Comment by vivekkiran** [comment]
![screenshot_2016-11-30-18-47-19-166](https://cloud.githubusercontent.com/assets/5557539/20753782/0afcf8b4-b72e-11e6-8941-91023704b9e6.jpeg)

![screenshot_2016-11-30-18-47-06-272](https://cloud.githubusercontent.com/assets/5557539/20753793/15e516b2-b72e-11e6-8b27-130a1c36d32a.jpeg)

**Comment by Tolriq** [comment]
Can confirm the ratio, I was about to give up on Moshi when found this issue.
Parsing a 750K json went from 7 sec to 0.7 after disabling instant run. (Everything last from AS to Moshi)

Was anything found? Is there anything I can provide to help find the issue?

**Comment by swankjesse** [comment]
Nothing yet. If you’d like to investigate, please do!

**Comment by Tolriq** [comment]
Well I'm very new to Moshi, but will report if I found anything after my global functions tests. ATM nice small yet functional library thanks for it.

Any hint at where / what to investigate first as I'm a total noob for the annotations handling in Java. (I do not use any adapter with annotations but I suppose looking for them could be a probable cause).

**Comment by swankjesse** [comment]
I think you start with figuring out how to profile code on instant run. My guess is that something in instant run is slow and the equivalent thing in regular Android is fast. What is it? Might be annotation loading; might be something else.
